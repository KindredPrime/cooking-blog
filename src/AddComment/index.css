.AddComment {
  margin-bottom: 40px;
  text-align: center;
}

.AddComment button {
  display: block;
  margin-top: 5px;
  margin-left: auto;
  margin-right: auto;
}