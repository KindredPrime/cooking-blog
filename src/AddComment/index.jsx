import { Component } from 'react';
import PropTypes from 'prop-types';
import APIError from '../APIError/index';
import './index.css';

class AddComment extends Component {
  state = {
    content: '',
    error: null
  };

  handleSubmit(e) {
    e.preventDefault();

    this.props.handleSubmit(this.state.content)
      // Clear the comment text
      .then(() => this.setState({
        content: '',
        error: null
      }))
      .catch((error) => this.setState({
        error
      }));
  }

  render() {
    const { content, error } = this.state;

    return (
      <form className="AddComment" onSubmit={(e) => this.handleSubmit(e)}>
        <textarea
          id="content"
          name="content"
          value={content}
          onChange={(e) => this.setState({
            content: e.target.value
          })}
          required
        ></textarea>

        <button type="submit">
          Add Comment
        </button>

        {error && <APIError message={error.message} />}
      </form>
    );
  }
}

// handleSubmit must return a thenable object
AddComment.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

export default AddComment;