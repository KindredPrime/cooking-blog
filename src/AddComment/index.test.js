import ReactDOM from 'react-dom';
import { render, screen, waitFor } from '@testing-library/react';
import UserEvent from '@testing-library/user-event';
import AddComment from './index';

describe('AddComment Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <AddComment handleSubmit={() => {}} />,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders an error as expected', async () => {
    const error = 'Test Error';
    render(<AddComment handleSubmit={() => new Promise((resolve, reject) => {
      throw new Error(error);
    })} />);

    UserEvent.type(screen.getByRole('textbox'), 'New Comment');
    UserEvent.click(screen.getByText('Add Comment'));

    await waitFor(() => expect(screen.getByText(error)).toBeInTheDocument());
    expect(document.body).toMatchSnapshot();
  });
});