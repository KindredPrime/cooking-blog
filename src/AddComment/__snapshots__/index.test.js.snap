// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddComment Component renders an error as expected 1`] = `
<body>
  <div>
    <form
      class="AddComment"
    >
      <textarea
        id="content"
        name="content"
        required=""
      >
        New Comment
      </textarea>
      <button
        type="submit"
      >
        Add Comment
      </button>
      <p
        class="APIError"
      >
        Test Error
      </p>
    </form>
  </div>
</body>
`;
